tx-parser/src/lib.rs

```diff
@@ -30,181 +34,150 @@ fn fn_parser<O, I>(f: fn(I) -> ParseResult<O, I>, err: &'static str) -> TxParser
     parser(f).expected(err)
 }
 
-impl<I> Tx<I>
-    where I: Stream<Item = Value>
-{
-    fn integer() -> TxParser<i64, I> {
-        fn_parser(Tx::<I>::integer_, "integer")
-    }
+def_parser_fn!(Tx, integer, Value, i64, input, {
+    return satisfy_map(|x: Value| if let Value::Integer(y) = x {
+            Some(y)
+        } else {
+            None
+        })
+        .parse_stream(input);
+});
 
-    fn integer_(input: I) -> ParseResult<i64, I> {
-        return satisfy_map(|x: Value| if let Value::Integer(y) = x {
-                Some(y)
-            } else {
-                None
-            })
-            .parse_stream(input);
-    }
+// The above using a def_value_satisfy_parser_fn
+//def_value_satisfy_parser_fn!(Tx, integer, i64, if let Value::Integer(y) = x { Some(y) } else { None });
```

Comment:
> In practice this allows you to simply pass a function that accepts an `&edn::Value` and
> returns an `Option<$result_type>`: if a suitable value is at the front of the stream,
> it will be converted and returned by the parser; otherwise, the parse will fail.

So if you define a function:

```
fn value_to_i64(val: &Value) -> Option<i64> {
    if let &Value::Integer(i) = val { Some(i) } else { None }
}
```

you can use it here:

```
def_value_satisfy_parser_fn!(Tx, integer, i64, value_to_i64);
```

Now it turns out we already have that function: it's `Value::as_integer`. So this should work:


```
def_value_satisfy_parser_fn!(Tx, integer, i64, Value::as_integer);
```